src/action/listdir.rs

```diff
@@ -0,0 +1,732 @@
+/// Fills all fields of `Response` using path to the file.
+#[cfg(not(target_os = "linux"))]
+fn fill_response(file_path: &Path) -> Result<Response, Error> {
+    let metadata = fs::symlink_metadata(file_path)
+        .map_err(Error::ReadPath)?;
+
+    Ok(Response {
+        size: metadata.len(),
+        atime: get_access_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        path: file_path.clone().to_path_buf(),
+        crtime: get_creation_time(&metadata),
+        ..Default::default()
+    })
+}
+
+pub fn handle<S: Session>(session: &mut S, request: Request)
+                          -> session::Result<()> {
+    let dir_path = &request.path;
+    let mut paths: Vec<PathBuf> = dir_path.read_dir()
+        .map_err(Error::ReadPath)?.filter_map(|entry| entry.ok())
+        .map(|entry| entry.path()).collect();
+    paths.sort();
+
+    for file_path in &paths {
+        session.reply(fill_response(file_path)?)?;
+    }
+
+    Ok(())
+}
+
+/// Constructs `PathBuf` from `String`. If provided string is empty constructs
+/// `PathBuf` from "/".
+fn get_path(path: &Option<String>) -> PathBuf {
+    match path {
+        Some(path) if !path.is_empty() => PathBuf::from(path),
+        _ => PathBuf::from("/"),
+    }
+}
+
+/// Fills st_linux_flags field.
+#[cfg(target_os = "linux")]
+fn get_linux_flags(path: &Path) -> Option<u32> {
+    use std::os::raw::c_long;
+    use std::fs::File;
+    use std::os::unix::io::AsRawFd;
+
+    let file = match File::open(path) {
+        Ok(file) => file,
+        Err(_) => return None,
+    };
+    let mut linux_flags: c_long = 0;
+    let linux_flags_ptr: *mut c_long = &mut linux_flags;
+    unsafe {
+        match ioctls::fs_ioc_getflags(file.as_raw_fd(), linux_flags_ptr) {
+            0 => Some(linux_flags as u32),
+            _ => None,
+        }
+    }
+}
+
+impl super::Request for Request {
+
+    type Proto = ListDirRequest;
+
+    fn from_proto(proto: Self::Proto) -> Result<Request, session::ParseError> {
+        let missing = session::MissingFieldError::new;
+        let pathspec = proto.pathspec.ok_or(missing("path spec"))?;
+        let path_type = pathspec.pathtype
+            .ok_or(missing("path type"))?;
+        if path_type != PathType::Os as i32 {
+            return Err(ParseError::unsupported_value(
+                "path type", path_type.to_string()).into());
+        }
+        let path_option = pathspec.path_options
+            .unwrap_or(Options::CaseLiteral as i32);
+        if path_option != Options::CaseLiteral as i32 {
+            return Err(ParseError::unsupported_value(
+                "path option", path_option.to_string()).into());
+        };
+        Ok(Request {
+            path: get_path(&pathspec.path),
+        })
+    }
+}
+
+/// Converts idiomatic `SystemTime` to `u64` for the protocol buffer.
+fn get_time_since_unix_epoch(sys_time: &Option<SystemTime>) -> Option<u64> {
+    match sys_time {
+        Some(time) => match micros(time.clone()) {
```

Comment:
You can use `Some(ref time)` pattern, this way you don't need to use `.clone()`.

Reply:
Actually, `SystemTime` implements `Copy`, so you can just dereference it (`*time`, rather than `time.clone()`).